openmc/surface.py

```diff
@@ -936,33 +978,139 @@ class Cylinder(Surface):
         Type of the surface
 
     """
-    def __init__(self, r=1., boundary_type='transmission',
-                 name='', surface_id=None):
-        super().__init__(surface_id, boundary_type, name=name)
-        self.r = r
+    _type = 'cylinder'
+    _coeff_keys = ('x0', 'y0', 'z0', 'r', 'dx', 'dy','dz')
+
+    def __init__(self, x0=0., y0=0., z0=0., r=1., dx=0., dy=0., dz=1., **kwargs):
+
+        super().__init__(**kwargs)
+
+        for key, val in zip(self._coeff_keys, (x0, y0, z0, r, dx, dy, dz)):
+            setattr(self, key, val)
+
+    @property
+    def x0(self):
+        return self.coefficients['x0']
+
+    @property
+    def y0(self):
+        return self.coefficients['y0']
+
+    @property
+    def z0(self):
+        return self.coefficients['z0']
 
     @property
     def r(self):
         return self.coefficients['r']
 
+    @property
+    def dx(self):
+        return self.coefficients['dx']
+
+    @property
+    def dy(self):
+        return self.coefficients['dy']
+
+    @property
+    def dz(self):
+        return self.coefficients['dz']
+
+    @x0.setter
+    def x0(self, x0):
+        check_type('x0 coefficient', x0, Real)
+        self._coefficients['x0'] = x0
+
+    @y0.setter
+    def y0(self, y0):
+        check_type('y0 coefficient', y0, Real)
+        self._coefficients['y0'] = y0
+
+    @z0.setter
+    def z0(self, z0):
+        check_type('z0 coefficient', z0, Real)
+        self._coefficients['z0'] = z0
+
     @r.setter
     def r(self, r):
         check_type('r coefficient', r, Real)
         self._coefficients['r'] = r
 
+    @dx.setter
+    def dx(self, dx):
+        check_type('dx coefficient', dx, Real)
+        self._coefficients['dx'] = dx
+
+    @dy.setter
+    def dy(self, dy):
+        check_type('dy coefficient', dy, Real)
+        self._coefficients['dy'] = dy
+
+    @dz.setter
+    def dz(self, dz):
+        check_type('dz coefficient', dz, Real)
+        self._coefficients['dz'] = dz
+
+    def _get_base_coeffs(self):
+        x0, y0, z0, r = self.x0, self.y0, self.z0, self.r
+        dx, dy, dz = self.dx, self.dy, self.dz
+        dx2, dy2, dz2 = dx*dx, dy*dy, dz*dz
+
+        a = dy2 + dz2
+        b = dx2 + dz2
+        c = dx2 + dy2
+        d = -2*dx*dy
+        e = -2*dy*dz
+        f = -2*dx*dz
+        g = 2*(dx*(z0*dz + y0*dy) - x0*(dy2 + dz2))
+        h = 2*(dy*(z0*dz + x0*dx) - y0*(dx2 + dz2))
+        j = 2*(dz*(y0*dy + x0*dx) - z0*(dx2 + dy2))
+        k = x0*x0*(dy2 + dz2) + y0*y0*(dx2 + dz2) + z0*z0*(dx2 + dy2) \
+            + e*y0*z0 + f*x0*z0 + d*x0*y0 - r*r*(dx2 + dy2 + dz2)
```

Comment:
Not at all obvious how one gets these coefficients (_I know_, but someone else coming in and looking at this would probably have no idea). I would suggest copying the comment from the `cylinder_from_points` function in `openmc/model/funcs.py` that describes how these are obtained.

Also, as you'll see there, I have a slightly simpler way of coming up with the g, h, j, k coefficients if you're interested in adopting that here. I confirmed that what you have is equivalent to what's there (thanks to sympy).

Reply:
Ok I just copied the cylinder from points method into this section, thanks!